scripts/release/generate_release_trigger_upload_jobs.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import sys
+
+from ros_buildfarm.argument import add_argument_config_url
+from ros_buildfarm.argument import add_argument_dry_run
+from ros_buildfarm.common import JobValidationError
+from ros_buildfarm.common import get_release_job_prefix
+from ros_buildfarm.config import get_index
+from ros_buildfarm.config import get_release_build_files
+from ros_buildfarm.jenkins import configure_job
+from ros_buildfarm.jenkins import connect
+from ros_buildfarm.templates import expand_template
+
+
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description="Generate the 'upload_main' and 'upload_testing' jobs.")
+    add_argument_config_url(parser)
+    add_argument_dry_run(parser)
+    args = parser.parse_args(argv)
+
+    template_name = 'release/trigger_upload_repo_job.xml.em'
+
+    config = get_index(args.config_url)
+    jenkins = connect(config.jenkins_url)
+
+    for repo in ['main', 'testing']:
+        job_name = 'upload_%s' % repo
+        block_when_upstream_building = 'true'
+        if repo == 'testing':
+            block_when_upstream_building = 'false'
+        job_config = expand_template(template_name, {
+            'block_when_upstream_building': block_when_upstream_building,
+            'repo': repo,
+            'upstream_job_names': get_upstream_job_names(config, repo)})
+
+        configure_job(jenkins, job_name, job_config, dry_run=args.dry_run)
+
+
+def get_upstream_job_names(config, repo):
+    distributions = config.distributions.keys()
+    if repo == 'main':
+        upstream_job_names = ['{0}_sync-packages-to-{1}'.format(
+            get_release_job_prefix(rosdistro), repo) for rosdistro in distributions]
+    elif repo == 'testing':
+        upstream_job_names = []
+        for rosdistro in distributions:
```

Comment:
Instead of creating the variable `distributions` this should use `config.distributions.keys()` directly. Otherwise it is not obvious that the `if` block as well as the `elif` block use the same generator.

Also replace the usage a few lines above and then remove the variable `distributions`.

Reply:
To be clear, your preference here is that the two cases do not share work if they are able?

Reply:
The reuse of the same generator seems to be fine (I just had to convince myself in an interactive Python shell).

This can stay as-i. Sorry for the noise.